Support stopping Interactive Tools without deleting their outputs

natefoo opened the pull request:
This is done by adding new `STOP` and `STOPPED` job states that work similar to (and use the same job stop queue in the job handler) `DELETED_NEW` and `DELETED`. Unlike deleted jobs, stopped jobs are expected to run through the normal job wrapper finish method after the job is killed.

Caveats:
- Only possible from the Active GxIT list currently (idea: for HDAs, can we make the delete "X" a stop sign on GxIT outputs, and once the user clicks that and the job is stopped, it becomes the regular "X"?)
- External metadata will never run (~~no idea what to do about this~~ added an option to retry *just GxIT* set_meta internally, which isn't perfect, but it's better than nothing).
- ~~Local and DRMAA runners only (so far - will do the others before this PR gets out of draft)~~ done except Chronos and probably Pulsar MQ.

natefoo committed:
Support stopping Interactive Tools without deleting their outputs by
adding new "STOP" and "STOPPED" job states - these are expected to run
through the normal job wrapper finish method after the job is killed.

- Only possible from the Active GxIT list currently
- External metadata will never run
- Local and DRMAA runners only (so far)

natefoo committed:
Add option to retry setting GxIT metadata internally.

natefoo committed:
Update job runners for STOPPED job state

nsoranzo commented on lib/galaxy/jobs/handler.py:
```diff
@@ -200,7 +200,8 @@ def __check_jobs_at_startup(self):
             in_list = (model.Job.states.QUEUED,
-                       model.Job.states.RUNNING)
+                       model.Job.states.RUNNING,
+                       model.Job.states.STOPPED)
```
Need to add 'stopped' to the method docstring?

nsoranzo commented on lib/galaxy/jobs/runners/cli.py:
```diff
@@ -168,12 +168,13 @@ def check_watched_items(self):
                 state = job_interface.parse_single_status(cmd_out.stdout, external_job_id)
                 if not state == model.Job.states.OK:
                     log.warning(f'({id_tag}/{external_job_id}) job not found in batch state check, but found in individual state check')
+            job_state = ajs.job_wrapper.get_state()
```
You can move this to line 160 and re-use it on line 161.

natefoo replied:
This was actually intentional as the single status check in the intervening lines could take a while, so I was worried the status might be stale by here if it was fetched earlier.

nsoranzo commented on lib/galaxy/model/__init__.py:
```diff
@@ -687,7 +687,9 @@ class Job(JobLike, UsesCreateAndUpdateTime, Dictifiable, RepresentById):
                    DELETED='deleted',
-                   DELETED_NEW='deleted_new')
+                   DELETED_NEW='deleted_new',
+                   STOP='stop',
```
As mentioned on Gitter, STOPPING seems to me more appropriate for a job state, even if it is a request to stop the job.

natefoo replied:
Will do. Thanks for the review!

natefoo committed:
Update docstring as suggested in code review

natefoo committed:
Rename job states DELETED_NEW -> DELETING and STOP -> STOPPING

jmchilton commented on lib/galaxy/jobs/__init__.py:
```diff
@@ -1536,6 +1536,8 @@ def _finish_dataset(self, output_name, dataset, job, context, final_job_state, r
             # it would be quicker to just copy the metadata from the originating output dataset,
             # but somewhat trickier (need to recurse up the copied_from tree), for now we'll call set_meta()
             retry_internally = util.asbool(self.get_destination_configuration("retry_metadata_internally", True))
+            if self.tool.tool_type == 'interactive':
```
``not retry_internally and self.tool.tool_type == 'interactive'``

mvdbeek committed:
Add testcase and move stop operation into API

mvdbeek committed:
Adjust client to new API endpoint

mvdbeek committed:
Respect retry_internally if it is already set

mvdbeek committed:
Adjust API test to new deleting/deleted states

mvdbeek committed:
Set exit_code for stopped job to 0

Pulsar sets the exit code for a stopped job to -9 otherwise.
Should probably implement some special handling for this.

mvdbeek committed:
Fix typo

mvdbeek reviewed (approved):
@natefoo can you check the changes I've made ? Especially the logic with ignoring the pulsar returncode.
